crates/bifrost/src/loglets/local_loglet/mod.rs

```diff
@@ -0,0 +1,202 @@
+            seal,
+            release_watch: OffsetWatch::new(release_pointer),
+        };
+        debug!(
+            log_id = log_id,
+            release_pointer = %release_pointer,
+            next_offset = loglet.next_write_offset.load(Ordering::Relaxed),
+            "Local loglet started"
+        );
+
+        Ok(loglet)
+    }
+
+    #[inline]
+    fn notify_readers(&self) {
+        let release_pointer = LogletOffset(self.last_committed_offset.load(Ordering::Relaxed));
+        self.release_watch.notify(release_pointer);
+    }
+
+    fn read_after(&self, after: LogletOffset) -> Result<Option<LogRecord<LogletOffset>>, Error> {
+        let trim_point = LogletOffset(self.trim_point_offset.load(Ordering::Relaxed));
+        // Are we reading after before the trim point? Note that if `trim_point` == `after`
+        // then we don't return a trim gap, because the next record is potentially a data
+        // record.
+        if trim_point > after {
+            return Ok(Some(LogRecord::new_trim_gap(after.next(), trim_point)));
+        }
+
+        let from_offset = after.next();
+        // Are we reading after commit offset?
+        let commit_offset = LogletOffset(self.last_committed_offset.load(Ordering::Relaxed));
+        if from_offset > commit_offset {
+            Ok(None)
+        } else {
+            let key = RecordKey::new(self.log_id, from_offset);
+            let data_cf = self.log_store.data_cf();
+            let mut read_opts = rocksdb::ReadOptions::default();
+            read_opts.set_iterate_upper_bound(RecordKey::upper_bound(self.log_id).to_bytes());
+
+            let mut iter = self.log_store.db().iterator_cf_opt(
+                data_cf,
+                read_opts,
+                rocksdb::IteratorMode::From(&key.to_bytes(), rocksdb::Direction::Forward),
+            );
+            let record = iter.next().transpose().map_err(LogStoreError::Rocksdb)?;
+            let Some(record) = record else {
+                return Ok(None);
+            };
+
+            let (key, data) = record;
+            let key = RecordKey::from_slice(&key);
+            // Defensive, the upper_bound set on the iterator should prevent this.
+            if key.log_id != self.log_id {
+                warn!(
+                    log_id = self.log_id,
+                    "read_after moved to the adjacent log {}, that should not happen.\
+                    This is harmless but needs to be investigated!",
+                    key.log_id,
+                );
+                return Ok(None);
+            }
+            let data = Bytes::from(data);
+            Ok(Some(LogRecord::new_data(key.offset, Payload::from(data))))
+        }
+    }
+}
+
+#[async_trait]
+impl LogletBase for LocalLoglet {
+    type Offset = LogletOffset;
+    async fn append(&self, payload: Payload) -> Result<LogletOffset, Error> {
+        let offset = LogletOffset::from(self.next_write_offset.fetch_add(1, Ordering::Relaxed));
```

Comment:
Can it happen that we might skip reading a record in rare cases because the log writer client decides on the write offset and not the log writer server? 

I have the following scenario in mind: Let's assume we have two writers `W1` and `W2` and a reader `R`. `R` wants to read the next record after lsn `0`. Now `W1` and `W2` append both a record. `W1` sends `P1 = Put(1, Pay1)` and `W2` sends `P2 = Put(2, Pay2)`. Now because of some unfortunate thread interleavings, `P2` arrives first at the log writer. Because of a bit more bad luck, the writer closes the batch and commits it immediately after processing `P2`. Now `W2` would set the commit offset to `2` and notify all waiting readers. `R` would then try reading after `0` which should return `Pay2` because `P1` hasn't been processed yet.

If this can happen, then it might help move the write offset decision/sequencer into the `RocksDbLogWriter`.

Reply:
You are absolutely right. Great catch. I will fix.